service/controller/clusterapi/v16/resources/awsclusterconfig/create.go

```diff
@@ -21,7 +28,125 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	machineDeployments, err := r.getMachineDeployments(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var versionBundles []versionbundle.Bundle
+	{
+		req := searcher.Request{
+			ReleaseVersion: key.ClusterReleaseVersion(cluster),
+		}
+
+		res, err := r.clusterClient.Release.Searcher.Search(ctx, req)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		versionBundles = res.VersionBundles
+	}
+
+	// Get existing AWSClusterConfig or create a new one.
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if AWSClusterConfig %q/%q exists", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+
+	awsClusterConfig, err := r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Get(key.AWSClusterConfigName(cluster), v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find AWSClusterConfig %q/%q", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating AWSClusterConfig %q/%q", awsClusterConfig.Namespace, key.AWSClusterConfigName(cluster)))
+
+		awsClusterConfig = r.constructAWSClusterConfig(cluster, machineDeployments, versionBundles)
+
+		_, err = r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Create(awsClusterConfig)
+		if errors.IsAlreadyExists(err) {
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("AWSClusterConfig %q/%q already exists", awsClusterConfig.Namespace, awsClusterConfig.Name))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created AWSClusterConfig %q/%q", awsClusterConfig.Namespace, awsClusterConfig.Name))
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found AWSClusterConfig %q/%q", cluster.Namespace, awsClusterConfig.Name))
+
+	// Map desired state from Cluster to AWSClusterConfig.
+	r.mapClusterToAWSClusterConfig(awsClusterConfig, cluster, machineDeployments, versionBundles)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating AWSClusterConfig %q/%q", awsClusterConfig.Namespace, awsClusterConfig.Name))
+
+	_, err = r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Update(awsClusterConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated AWSClusterConfig %q/%q", awsClusterConfig.Namespace, awsClusterConfig.Name))
 
 	return nil
 }
+
+func (r *Resource) getMachineDeployments(ctx context.Context, cluster clusterv1alpha1.Cluster) ([]clusterv1alpha1.MachineDeployment, error) {
+	labelSelector := v1.AddLabelToSelector(&v1.LabelSelector{}, label.Cluster, key.ClusterID(cluster))
+	// TODO: Add selector for provider annotation?
```

Comment:
There should only ever be one provider at an installation. Why should we need a label for it?

Reply:
Logically there should only ever be one, but in case someone accidentally applies YAMLs to wrong installation by accidentally having wrong connection, it makes `cluster-operator` to immediately panic due to our approach in key package. We would need to remove panics from there. For me having the selector here would be the right thing as it doesn't have any drawbacks but makes it safer.